scripts/build-windows.ps1

```diff
@@ -18,28 +14,97 @@ $env:CSC_IDENTITY_AUTO_DISCOVERY = $false
 
 $env:CI = $true
 
-$EXPECTED_NODE_VERSION = (Get-Content .\.nvmrc)
-$ACTUAL_NODE_VERSION = (node -v)
+# -----------------------------------------------------------------------------
+#                  Utility functions
 
-if ( "v$EXPECTED_NODE_VERSION" -ne $ACTUAL_NODE_VERSION)
-{
-  Write-Host "You are not running the expected version of node!"
-  Write-Host "  expected: [v$EXPECTED_NODE_VERSION]"
-  Write-Host "  actual  : [$ACTUAL_NODE_VERSION]"
+Function fail_with_docs {Param ($1)
+  Write-Host "*************** FAILING ***************"
+  Write-Host "$1"
+  Write-Host ""
+  Write-Host "Please read the developer documentation in CONTRIBUTING.md"
   exit 1
 }
 
+Function Test-CommandExists { Param ($command, $1)
+
+  $oldPreference = $ErrorActionPreference
+  $ErrorActionPreference = "stop"
+
+  try {
+    if(Get-Command $command){RETURN}
+  } Catch {
+    fail_with_docs $1
+  }
+  Finally {$ErrorActionPreference=$oldPreference}
+}
+
+# -----------------------------------------------------------------------------
+#                  Checking the developer environment
+# checking for installed programmes
+Test-CommandExists node "Node is not installed"
+Test-CommandExists npm "npm is not installed"
+Test-CommandExists python "Python is not installed"
+# NEEDS proper way to CHECK MSVS Tools
+
+# Checking node.js version
+$EXPECTED_NODE_VERSION = (cat .nvmrc)
+$ACTUAL_NODE_VERSION = (node -v)
+
+if ( "v$EXPECTED_NODE_VERSION" -ne $ACTUAL_NODE_VERSION) {
+  fail_with_docs "You are not running the expected version of node!
+   expected: [v$EXPECTED_NODE_VERSION]
+   actual  : [$ACTUAL_NODE_VERSION]"
+}
+
+# Checking npm version
+$EXPECTED_NPM_VERSION = (Get-Content package.json | ConvertFrom-Json).engines.npm 
+$ACTUAL_NPM_VERSION = (npm -v)
+
+if ( $EXPECTED_NPM_VERSION -ne $ACTUAL_NPM_VERSION) {
+  Write-Host "You are not running the expected version of npm!
+   expected: [$EXPECTED_NPM_VERSION]
+   actual  : [$ACTUAL_NPM_VERSION]"
+  Write-Host "Changing version of to [$EXPECTED_NPM_VERSION]"
+}
+
+# Checking python version
+$EXPECTED_PYTHON_VERSION = "3.6"
```

Comment:
python version being discussed earlier in the thread was `3.10.4`

Reply:
Not sure if is in fact a dependency. I looked for the image I posted on discord and python 3.6 was the lowest possible on the error I got. Shall I keep 3.6 or force at least 3.7?

Reply:
![unknown](https://user-images.githubusercontent.com/37463445/164784177-6f115bf8-1199-4595-925e-23f000038764.png)

Reply:
i think the `CONTRIBUTING.md` has `3.10.4` - which works on my machine (macos)

Reply:
Version referenced in `CONTRIBUTING.md` is python 3.10.4, so we should update here or in `CONTRIBUTING.md` with a version that works :-)

Reply:
Yeah, it was me who added that on CONTRIBUTING.md . I can try to build everything with python 3.6 and I will tell you if it worked or not, so as to find that if it is a requirement or not.

What do you think about this?

Reply:
yes, we can go with whatever works across all 3 OSes that we usually build for. If its the lower version of python, im fine to downgrade - not a problem. Just that we need to be consistent across all 3 OSes so that we can uncover similar issues

Reply:
in the latest commit (for now), I have set the python version to `3.10.4`. please confirm if we want to go with this or the lower one. tx

Reply:
The expression that I have on the code implies 3.6 or higher, which means that it would work with whatever version that is higher, so it wouldn't break if we are using the latest python.

Reply:
yes, but as i had mentioned in other threads, im not in favor of keeping it open-ended, there will be issues in the future where things break randomly on different machines due to versions being different. This has happened in the past, and I prefer to remain locked. The upgrade needs to be an explicit intentional action from a dev and then it can be monitored for all OSes.

Reply:
Resolving for now - we can touch back on this specific topic in a later PR/thread